File: editor/input.go
package editor

// InputState tracks multi-key sequence state
type InputState struct {
	// Numeric prefix accumulator
	Count    int
	HasCount bool

	// Pending character find
	PendingFindForward  bool
	PendingFindBackward bool

	// Pending goto line (after typing :)
	PendingGotoLine bool
	GotoLineBuffer  string

	// Last find character for ; and , repeat
	LastFindChar    rune
	LastFindForward bool
	HasLastFind     bool

	// Pending operator for commands like dd, yy, dw, etc.
	PendingOperator rune

	// Search mode state
	Search *SearchState

	// Autocomplete state (nil when not active)
	Autocomplete *AutocompleteState

	// Pending mark operations
	PendingMark       bool // True when waiting for mark identifier after 'm'
	PendingJumpToMark bool // True when waiting for mark identifier after '`'
}

// NewInputState creates a new input state
func NewInputState() *InputState {
	return &InputState{}
}

// Reset clears all pending state except last find memory
func (s *InputState) Reset() {
	s.Count = 0
	s.HasCount = false
	s.PendingFindForward = false
	s.PendingFindBackward = false
	s.PendingGotoLine = false
	s.GotoLineBuffer = ""
	s.PendingOperator = 0
	s.PendingMark = false
	s.PendingJumpToMark = false
}

// GetCount returns the count, defaulting to 1 if not set
func (s *InputState) GetCount() int {
	if s.HasCount {
		return s.Count
	}
	return 1
}

// AddDigit appends a digit to the count
func (s *InputState) AddDigit(d int) {
	s.Count = s.Count*10 + d
	s.HasCount = true
}

// HasPending returns true if any pending state is active
func (s *InputState) HasPending() bool {
	return s.PendingFindForward || s.PendingFindBackward || s.PendingGotoLine || s.PendingOperator != 0 || s.PendingMark || s.PendingJumpToMark
}

// PendingString returns a string representation of pending input
func (s *InputState) PendingString() string {
	if s.PendingFindForward {
		return "f_"
	}
	if s.PendingFindBackward {
		return "F_"
	}
	if s.PendingGotoLine {
		return ":" + s.GotoLineBuffer
	}
	if s.PendingOperator != 0 {
		return string(s.PendingOperator)
	}
	if s.PendingMark {
		return "m_"
	}
	if s.PendingJumpToMark {
		return "`_"
	}
	return ""
}

// SaveLastFind stores the last find character and direction
func (s *InputState) SaveLastFind(ch rune, forward bool) {
	s.LastFindChar = ch
	s.LastFindForward = forward
	s.HasLastFind = true
}

File: c_app/src/input.c
#include "input.h"
#include "autocomplete.h"

#include <stdlib.h>
#include <string.h>

InputState *input_state_new(void) {
    InputState *s = calloc(1, sizeof(InputState));
    return s;
}

void input_state_free(InputState *s) {
    if (!s) return;
    ac_state_free(s->autocomplete);
    free(s);
}

void input_state_reset(InputState *s) {
    s->count = 0;
    s->has_count = false;
    s->pending_find_forward = false;
    s->pending_find_backward = false;
    s->pending_goto_line = false;
    s->goto_line_buffer[0] = '\0';
    s->goto_line_buf_len = 0;
    s->pending_operator = 0;
    s->pending_mark = false;
    s->pending_jump_to_mark = false;
}

int input_state_get_count(const InputState *s) {
    if (s->has_count) return s->count;
    return 1;
}

void input_state_add_digit(InputState *s, int d) {
    s->count = s->count * 10 + d;
    s->has_count = true;
}

bool input_state_has_pending(const InputState *s) {
    return s->pending_find_forward || s->pending_find_backward
        || s->pending_goto_line || s->pending_operator != 0
        || s->pending_mark || s->pending_jump_to_mark;
}

void input_state_save_last_find(InputState *s, wchar_t ch, bool forward) {
    s->last_find_char = ch;
    s->last_find_forward = forward;
    s->has_last_find = true;
}
